src/main/java/org/gridsuite/modification/server/modifications/GeneratorCreation.java

```diff
@@ -201,4 +206,264 @@ private void createGeneratorInBusBreaker(VoltageLevel voltageLevel, GeneratorCre
+                generatorCreationInfos.getRegulatingTerminalType() != null) {
+                Terminal terminal = ModificationUtils.getInstance().getTerminalFromIdentifiable(voltageLevel.getNetwork(),
+                    generatorCreationInfos.getRegulatingTerminalId(),
+                    generatorCreationInfos.getRegulatingTerminalType(),
+                    generatorCreationInfos.getRegulatingTerminalVlId());
+                if (terminal != null) {
+                    updateGeneratorRegulatingTerminal(generatorCreationInfos, generator, terminal, voltageReports);
+                }
+            }
+            if (generatorCreationInfos.getQPercent() != null) {
+                try {
+                    generator.newExtension(CoordinatedReactiveControlAdderImpl.class)
+                        .withQPercent(generatorCreationInfos.getQPercent()).add();
+                    voltageReports.add(ModificationUtils.getInstance().buildCreationReport(generatorCreationInfos.getQPercent(), "Reactive percentage"));
+                } catch (PowsyblException e) {
+                    voltageReports.add(Report.builder()
+                        .withKey("ReactivePercentageError")
+                        .withDefaultMessage("cannot add Coordinated reactive extension on generator with id=${id} :" + e.getMessage())
+                        .withValue("id", generatorCreationInfos.getEquipmentId())
+                        .withSeverity(TypedValue.ERROR_SEVERITY)
+                        .build());
+                }
+            }
+            ModificationUtils.getInstance().reportModifications(subReporter, voltageReports, "VoltageRegulationCreated", "Voltage regulation");
+        }
+    }
+
+    private void updateGeneratorRegulatingTerminal(GeneratorCreationInfos generatorCreationInfos, Generator generator,
+                                                   Terminal terminal, List<Report> voltageReports) {
+        if (generatorCreationInfos.getRegulatingTerminalId() != null
+                && generatorCreationInfos.getRegulatingTerminalType() != null
+                && generatorCreationInfos.getRegulatingTerminalVlId() != null) {
+            generator.setRegulatingTerminal(terminal);
+            voltageReports.add(ModificationUtils.getInstance().buildCreationReport(
+                    generatorCreationInfos.getRegulatingTerminalVlId(),
+                    "Voltage level"));
+            voltageReports.add(ModificationUtils.getInstance().buildCreationReport(
+                    generatorCreationInfos.getRegulatingTerminalType() + ":"
+                            + generatorCreationInfos.getRegulatingTerminalId(),
+                    "Equipment"));
+        }
+    }
+
+    private void reportGeneratorConnection(GeneratorCreationInfos generatorCreationInfos, Reporter subReporter) {
+        if ((generatorCreationInfos.getVoltageLevelId() != null && generatorCreationInfos.getBusOrBusbarSectionId() != null)
```

Comment:
From my heretic way of writing, I would replace the first `&&` condition by a
```
if (generatorCreationInfos.getVoltageLevelId() == null) {
  return;
}
if (generatorCreationInfos.getBusOrBusbarSectionId() == null) {
  return;
}
```
and then, after the three possible additions,
```
if (!connectionReports.isEmpty()) {
  ModificationUtils.getInstance().reportModifications(subReporter, connectionReports, "ConnectionCreated", CONNECTION);
}
```
May be factorize `ModificationUtils.getInstance()`

Reply:
Done